File: Java/src/World/body/Plant.java
package World.body;

import GUI.Color_obj;
import World.Point;
import World.World;

import java.util.concurrent.ThreadLocalRandom;

public abstract class Plant extends Body{

    public Plant(Point point_location, World world) {
        super(point_location, world);
        setInitiative(0);
    }

    @Override
    public void move() {
        action();
    }
    public abstract int get_stronger();
    @Override
    protected void action() {
        int randomNumber = ThreadLocalRandom.current().nextInt(1, 101);
        if(randomNumber<7){
            int free_spaces = getWorld().free_spaces(this);
            if(free_spaces>0){
                Point new_point = new Point(getPoint_location().getX(), getPoint_location().getY());
                random_location_empty(this, new_point);
                new_body(new_point);
                getWorld().getNews_panel().add_news(getName() + " has grown on (" + new_point.getX() + ", " + new_point.getY()+')');
            }
            else{
                getWorld().getNews_panel().add_news("Nothing happened");
            }
        }
        else{
            getWorld().getNews_panel().add_news("Nothing happened");
        }
    }

    @Override
    protected void collision(Body attacker) {
        if(this.getPower()>attacker.getPower()){
            getWorld().getNews_panel().add_news(attacker.getName() + " has eaten " + getName() + " and dies on (" + attacker.getPoint_location().getX() + ", " + attacker.getPoint_location().getY()+')');
            getWorld().getBoxes()[attacker.getPoint_location().getY()-1][attacker.getPoint_location().getX()-1].setColor(Color_obj.EMPTY);
            getWorld().delete_body2(attacker);
            getWorld().delete_body(this);
        }
        else{
            getWorld().getNews_panel().add_news(attacker.getName() + " has eaten " + getName() + " on (" + getPoint_location().getX() + ", " + getPoint_location().getY()+')');
            if(get_stronger()>0){
                attacker.setPower(attacker.getPower()+get_stronger());
                getWorld().getNews_panel().add_news(attacker.getName() + " has grown stronger and now has " + attacker.getPower() + " power");
            }
            getWorld().delete_body(this);
            getWorld().getBoxes()[getPoint_location().getY()-1][getPoint_location().getX()-1].setColor(attacker.getColor());
        }
    }


}

File: Python/World/Body/plant.py
from abc import abstractmethod, ABC
import random

from GUI.colors import Colors
from World.Body.body import Body
from World.point import Point


class Plant(Body, ABC):
    def __init__(self, point_location, world):
        super().__init__(point_location, world)
        self.set_initiative(0)

    def move(self):
        self.action()

    def action(self):
        randomNumber = random.randint(1, 100)
        if randomNumber < 7:
            free_spaces = self.world.free_spaces(self)
            if free_spaces > 0:
                new_point = Point(self.get_point_location().get_x(), self.get_point_location().get_y())
                self.random_location_empty(self, new_point)
                self.new_body(new_point)
                self.world.mainWindow.add_text(self.get_name() + " has grown on (" + str(new_point.get_x()) + ", " + str(new_point.get_y()) + ")")
            else:
                self.world.mainWindow.add_text("Nothing happened")
        else:
            self.world.mainWindow.add_text("Nothing happened")

    def collision(self, attacker):
        if self.get_power() > attacker.get_power():
            self.world.mainWindow.add_text(attacker.get_name() + " has eaten " + self.get_name() +
                    "and dies on (" + str(self.get_point_location().get_x()) + ", " + str(self.get_point_location().get_y()) + ")")
            self.world.get_boxes[attacker.get_point_location().get_y() - 1][
                attacker.get_point_location().get_x() - 1].set_color(Colors.EMPTY)
            self.world.delete_body2(attacker)
            self.world.delete_body(self)
        else:
            self.world.mainWindow.add_text(attacker.get_name() + " has eaten " + self.get_name() +
                            "on ("+ str(self.get_point_location().get_x()) + ", " + str(self.get_point_location().get_y()) + ")")
            if self.get_stronger() > 0:
                attacker.set_power(attacker.get_power() + self.get_stronger())
                self.world.mainWindow.add_text(attacker.get_name() + " has grown stronger and now has " + str(attacker.get_power()) + " power")
            self.world.delete_body(self)
            self.world.get_boxes[self.get_point_location().get_y() - 1][
                self.get_point_location().get_x() - 1].set_color(attacker.get_color())

    @abstractmethod
    def get_stronger(self):
        pass
